File: scripts/genmask.go
//go:build ignore

// genmask generates the 3600×1800 bit-packed land mask asset.
//
// Usage:
//
//	go run scripts/genmask.go
//
// Output: internal/maprender/assets/landmask.bin
// Format: 3600*1800 bits (810 000 bytes), MSB-first within each byte.
// Row-major: lat 90→-90 (rows 0-1799), lon -180→180 (cols 0-3599), 0.1°/cell.
//
// Data source: Natural Earth 10m land polygons (ne_10m_land.geojson, ~10MB).
// No external dependencies — stdlib only.
package main

import (
	"encoding/json"
	"log"
	"net/http"
	"os"
	"runtime"
	"sync"
	"time"
)

const (
	cols   = 3600
	rows   = 1800
	url    = "https://raw.githubusercontent.com/nvkelso/natural-earth-vector/master/geojson/ne_10m_land.geojson"
	outDir = "internal/maprender/assets"
	outFile = outDir + "/landmask.bin"
)

// ring is a closed polygon ring stored flat as alternating [lon, lat] pairs.
type ring struct {
	pts          [][2]float64
	minX, maxX   float64
	minY, maxY   float64
}

type polygon struct {
	exterior ring
	holes    []ring
}

func newRing(coords []interface{}) ring {
	r := ring{pts: make([][2]float64, len(coords))}
	r.minX, r.minY = 1e9, 1e9
	r.maxX, r.maxY = -1e9, -1e9
	for i, p := range coords {
		pt := p.([]interface{})
		x := pt[0].(float64)
		y := pt[1].(float64)
		r.pts[i] = [2]float64{x, y}
		if x < r.minX {
			r.minX = x
		}
		if x > r.maxX {
			r.maxX = x
		}
		if y < r.minY {
			r.minY = y
		}
		if y > r.maxY {
			r.maxY = y
		}
	}
	return r
}

func (r ring) contains(px, py float64) bool {
	inside := false
	pts := r.pts
	n := len(pts)
	j := n - 1
	for i := 0; i < n; i++ {
		xi, yi := pts[i][0], pts[i][1]
		xj, yj := pts[j][0], pts[j][1]
		if (yi > py) != (yj > py) {
			if px < (xj-xi)*(py-yi)/(yj-yi)+xi {
				inside = !inside
			}
		}
		j = i
	}
	return inside
}

func (p polygon) contains(lon, lat float64) bool {
	if !p.exterior.contains(lon, lat) {
		return false
	}
	for _, h := range p.holes {
		if h.contains(lon, lat) {
			return false
		}
	}
	return true
}

// spatialIndex maps 10°×10° grid cell → polygons whose exterior bbox overlaps it.
type spatialIndex [36][18][]polygon

func (si *spatialIndex) add(p polygon) {
	ext := p.exterior
	gx0 := int((ext.minX + 180) / 10)
	gx1 := int((ext.maxX + 180) / 10)
	gy0 := int((90 - ext.maxY) / 10)
	gy1 := int((90 - ext.minY) / 10)
	if gx0 < 0 {
		gx0 = 0
	}
	if gx1 > 35 {
		gx1 = 35
	}
	if gy0 < 0 {
		gy0 = 0
	}
	if gy1 > 17 {
		gy1 = 17
	}
	for gy := gy0; gy <= gy1; gy++ {
		for gx := gx0; gx <= gx1; gx++ {
			si[gx][gy] = append(si[gx][gy], p)
		}
	}
}

func (si *spatialIndex) isLand(lon, lat float64) bool {
	gx := int((lon + 180) / 10)
	gy := int((90 - lat) / 10)
	if gx < 0 {
		gx = 0
	} else if gx > 35 {
		gx = 35
	}
	if gy < 0 {
		gy = 0
	} else if gy > 17 {
		gy = 17
	}
	for _, p := range si[gx][gy] {
		ext := p.exterior
		if lon < ext.minX || lon > ext.maxX || lat < ext.minY || lat > ext.maxY {
			continue
		}
		if p.contains(lon, lat) {
			return true
		}
	}
	return false
}

func fetchGeoJSON() ([]byte, error) {
	log.Printf("Downloading %s ...", url)
	resp, err := http.Get(url) //nolint:noctx
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	var buf []byte
	tmp := make([]byte, 1<<20)
	for {
		n, err := resp.Body.Read(tmp)
		if n > 0 {
			buf = append(buf, tmp[:n]...)
		}
		if err != nil {
			break
		}
	}
	log.Printf("  downloaded %d bytes", len(buf))
	return buf, nil
}

func buildIndex(data []byte) (*spatialIndex, error) {
	var geojson struct {
		Features []struct {
			Geometry struct {
				Type        string        `json:"type"`
				Coordinates []interface{} `json:"coordinates"`
			} `json:"geometry"`
		} `json:"features"`
	}
	if err := json.Unmarshal(data, &geojson); err != nil {
		return nil, err
	}

	idx := &spatialIndex{}
	totalPolys := 0
	for _, feat := range geojson.Features {
		g := feat.Geometry
		var polys [][]interface{}
		switch g.Type {
		case "Polygon":
			polys = [][]interface{}{g.Coordinates}
		case "MultiPolygon":
			for _, p := range g.Coordinates {
				polys = append(polys, p.([]interface{}))
			}
		}
		for _, rawPoly := range polys {
			poly := polygon{}
			for i, rawRing := range rawPoly {
				rr := rawRing.([]interface{})
				r := newRing(rr)
				if i == 0 {
					poly.exterior = r
				} else {
					poly.holes = append(poly.holes, r)
				}
			}
			idx.add(poly)
			totalPolys++
		}
	}
	log.Printf("  %d polygons indexed across %d features", totalPolys, len(geojson.Features))
	return idx, nil
}

func main() {
	data, err := fetchGeoJSON()
	if err != nil {
		log.Fatalf("download: %v", err)
	}

	log.Println("Parsing and building spatial index...")
	idx, err := buildIndex(data)
	if err != nil {
		log.Fatalf("build index: %v", err)
	}

	packed := make([]byte, rows*cols/8)
	nWorkers := runtime.NumCPU()
	log.Printf("Rasterizing %d×%d at %d workers...", cols, rows, nWorkers)

	rowCh := make(chan int, rows)
	for r := 0; r < rows; r++ {
		rowCh <- r
	}
	close(rowCh)

	var mu sync.Mutex
	var wg sync.WaitGroup
	t0 := time.Now()
	done := make([]int32, rows)

	for w := 0; w < nWorkers; w++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			rowBuf := make([]byte, cols/8)
			for r := range rowCh {
				lat := 89.95 - float64(r)*0.1
				clear(rowBuf)
				for c := 0; c < cols; c++ {
					lon := -179.95 + float64(c)*0.1
					if idx.isLand(lon, lat) {
						rowBuf[c/8] |= 1 << (7 - c%8)
					}
				}
				mu.Lock()
				copy(packed[r*cols/8:], rowBuf)
				done[r] = 1
				mu.Unlock()
			}
		}()
	}

	// Progress monitor
	go func() {
		for {
			time.Sleep(10 * time.Second)
			mu.Lock()
			n := 0
			for _, v := range done {
				n += int(v)
			}
			mu.Unlock()
			if n >= rows {
				return
			}
			elapsed := time.Since(t0)
			eta := time.Duration(float64(elapsed) / float64(max(n, 1)) * float64(rows-n))
			log.Printf("  %d/%d rows (%.0f%%)  elapsed %s  eta %s",
				n, rows, float64(n)/float64(rows)*100, elapsed.Round(time.Second), eta.Round(time.Second))
		}
	}()

	wg.Wait()

	if err := os.MkdirAll(outDir, 0o755); err != nil {
		log.Fatalf("mkdir: %v", err)
	}
	if err := os.WriteFile(outFile, packed, 0o644); err != nil {
		log.Fatalf("write: %v", err)
	}

	land := 0
	for _, b := range packed {
		land += int(b&1) + int((b>>1)&1) + int((b>>2)&1) + int((b>>3)&1) +
			int((b>>4)&1) + int((b>>5)&1) + int((b>>6)&1) + int((b>>7)&1)
	}
	total := rows * cols
	log.Printf("Written %d bytes to %s (%d land, %d ocean, %.1f%% land)",
		len(packed), outFile, land, total-land, float64(land)/float64(total)*100)
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: scripts/genmask.py
#!/usr/bin/env python3
"""Generate 3600x1800 bit-packed land mask from Natural Earth 10m land polygons.

Output: internal/maprender/assets/landmask.bin
Format: 3600*1800 bits (810000 bytes), MSB-first within each byte.
Row-major: lat 89.95->-89.95 (rows), lon -179.95->179.95 (cols), 0.1 deg/cell.

Preferred: use the Go version (no extra deps, parallel, ~seconds):
    go run scripts/genmask.go

Fast path (rasterio available):
    pip install rasterio fiona numpy
    python3 scripts/genmask.py

Pure-stdlib fallback (no extra packages needed, ~30 min):
    python3 scripts/genmask.py
"""

import json
import sys
import time
import urllib.request

URL = (
    "https://raw.githubusercontent.com/nvkelso/natural-earth-vector"
    "/master/geojson/ne_10m_land.geojson"
)
COLS = 3600
ROWS = 1800
OUT = "internal/maprender/assets/landmask.bin"

# ---------------------------------------------------------------------------
# Rasterio fast path
# ---------------------------------------------------------------------------

def _run_rasterio(geojson_path: str) -> bytearray:
    import numpy as np
    import fiona
    import rasterio
    from rasterio.features import rasterize
    from rasterio.transform import from_bounds

    transform = from_bounds(-180, -90, 180, 90, COLS, ROWS)
    with fiona.open(geojson_path) as src:
        shapes = [(f["geometry"], 1) for f in src]

    raw = rasterize(
        shapes,
        out_shape=(ROWS, COLS),
        transform=transform,
        fill=0,
        dtype=np.uint8,
    )

    # Bit-pack: MSB-first
    flat = raw.flatten()
    packed = bytearray(ROWS * COLS // 8)
    for idx, v in enumerate(flat):
        if v:
            packed[idx // 8] |= 1 << (7 - idx % 8)
    return packed


# ---------------------------------------------------------------------------
# Pure-stdlib fallback
# ---------------------------------------------------------------------------

def _bbox(ring: list) -> tuple:
    xs = [p[0] for p in ring]
    ys = [p[1] for p in ring]
    return min(xs), max(xs), min(ys), max(ys)


def _in_ring(px: float, py: float, ring: list) -> bool:
    inside = False
    j = len(ring) - 1
    for i in range(len(ring)):
        xi, yi = ring[i]
        xj, yj = ring[j]
        if ((yi > py) != (yj > py)) and (px < (xj - xi) * (py - yi) / (yj - yi) + xi):
            inside = not inside
        j = i
    return inside


def _in_poly(lon: float, lat: float, ext: list, holes: list) -> bool:
    return _in_ring(lon, lat, ext) and not any(_in_ring(lon, lat, h) for h in holes)


def _build_index(geometries: list) -> dict:
    """Build a 36x18 spatial grid (10-deg cells) mapping grid_key -> list of rings."""
    index: dict = {}
    for geom in geometries:
        gtype = geom["type"]
        polys = [geom["coordinates"]] if gtype == "Polygon" else geom["coordinates"]
        for poly in polys:
            ext = poly[0]
            holes = poly[1:]
            bx0, bx1, by0, by1 = _bbox(ext)
            # Expand to all 10-deg grid cells the bbox touches
            gx0 = max(int((bx0 + 180) // 10), 0)
            gx1 = min(int((bx1 + 180) // 10), 35)
            gy0 = max(int((90 - by1) // 10), 0)
            gy1 = min(int((90 - by0) // 10), 17)
            for gy in range(gy0, gy1 + 1):
                for gx in range(gx0, gx1 + 1):
                    index.setdefault((gx, gy), []).append((bx0, bx1, by0, by1, ext, holes))
    return index


def _run_stdlib(geojson_bytes: bytes) -> bytearray:
    data = json.loads(geojson_bytes)
    geometries = [f["geometry"] for f in data["features"]]
    print(f"  {len(geometries)} features loaded, building spatial index...", file=sys.stderr)

    index = _build_index(geometries)
    print(f"  index built ({len(index)} non-empty 10-deg grid cells)", file=sys.stderr)

    packed = bytearray(ROWS * COLS // 8)
    t0 = time.time()

    for row in range(ROWS):
        lat = 89.95 - row * 0.1
        gy = min(int((90 - lat) // 10), 17)

        for col in range(COLS):
            lon = -179.95 + col * 0.1
            gx = min(int((lon + 180) // 10), 35)
            rings = index.get((gx, gy), ())

            for bx0, bx1, by0, by1, ext, holes in rings:
                if bx0 <= lon <= bx1 and by0 <= lat <= by1:
                    if _in_poly(lon, lat, ext, holes):
                        idx = row * COLS + col
                        packed[idx // 8] |= 1 << (7 - idx % 8)
                        break

        if row % 180 == 0:
            elapsed = time.time() - t0
            pct = row / ROWS * 100
            eta = (elapsed / max(row, 1)) * (ROWS - row)
            print(
                f"  {pct:5.1f}%  row {row:4d}/{ROWS}  elapsed {elapsed:5.0f}s  eta {eta:5.0f}s",
                file=sys.stderr,
            )

    return packed


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main() -> None:
    print(f"Downloading {URL} ...", file=sys.stderr)
    with urllib.request.urlopen(URL) as resp:
        raw_bytes = resp.read()
    print(f"  downloaded {len(raw_bytes):,} bytes", file=sys.stderr)

    packed: bytearray

    try:
        import tempfile, os
        with tempfile.NamedTemporaryFile(suffix=".geojson", delete=False) as tmp:
            tmp.write(raw_bytes)
            tmp_path = tmp.name
        try:
            print("rasterio available — using fast path", file=sys.stderr)
            packed = _run_rasterio(tmp_path)
        finally:
            os.unlink(tmp_path)
    except ImportError:
        print("rasterio not available — using pure-Python path", file=sys.stderr)
        packed = _run_stdlib(raw_bytes)

    with open(OUT, "wb") as f:
        f.write(bytes(packed))

    land = sum(bin(b).count("1") for b in packed)
    total = ROWS * COLS
    print(
        f"Written {len(packed):,} bytes to {OUT}  "
        f"({land:,} land cells, {total - land:,} ocean, {land/total*100:.1f}% land)",
        file=sys.stderr,
    )


if __name__ == "__main__":
    main()
